### ISSUE TITLE
How to support fixed-size rvv intrinsic type in gcc ?

### ISSUE THREAD

**Issue by zengdage** [howjmay/neon2rvv#373]
Hi @howjmay ，I have a question about the neon intrinsic types convert into rvv intrinsic types. Can you help me with that ?

As I know，the neon intrinsic types are fixed-size，but rvv intrinsic types are sizeless. For example,  the size of uint8x16_t (neon type) is 128 bits，and the size of vuint8m1_t (rvv type) is unkown, but you bind them together in neon2rvv.h. 

I think it will make some errors in such scenario because the compiler need to allocate fixed-size stack memory space for vtmp.
```
uint8x16_t vtmp;
.......
vtmp = vld1q_u8(rp);
```

**Comment by howjmay** [comment]
Sorry I can't understand your question totally. 
I guess you may mention this case 
https://github.com/howjmay/neon2rvv/blob/main/tests/impl.cpp#L574
It should be covered in the test case

**Comment by zengdage** [comment]
Sorry, maybe I didn't explain it clearly and give a incomplete example. May be you can test the following code, 

```
#include <stdlib.h>

#if defined(__arm__) || defined(__aarch64__)
#include <arm_neon.h>
#else

#include <riscv_vector.h>

typedef vuint8m1_t uint8x16_t;

__attribute__((always_inline)) uint8x16_t vld1q_u8(const uint8_t *ptr) { return __riscv_vle8_v_u8m1(ptr, 16); }
__attribute__((always_inline)) void vst1q_u8(uint8_t *a, uint8x16_t b) { return __riscv_vse8_v_u8m1(a, b, 16); }
#endif

struct test_struct {
  uint8x16_t x;
};

int main(int argc, char **argv) {
  unsigned char *data;

  data = malloc(1024);
  if (!data) {
    return -1;
  }
  struct test_struct temp;
  temp.x = vld1q_u8(data);
  vst1q_u8(data + 512, temp.x);
  return 0;
}
```

Because the uint8x16_t is fixed-size in neon type system，so it can be a member in union or member, but vuint8x16_t can't.  I found that it have been supported in gcc and clang in https://github.com/riscv-non-isa/rvv-intrinsic-doc/issues/176.

**Comment by OMaghiarIMG** [comment]
Hello @howjmay 
Yes @zengdage is correct, if you want to use Neon2RVV in a project that uses Neon types in structs/global variables it needs fixed-size RVV types.
I could contribute the switch.

**Comment by mr-c** [comment]
The optimized RISCV64 implementations for Neon in SIMDe use a fixed size


> We have included clang-qemu-rvv testing for the following RISC-V V Extension architectures, both with and without ZVFH enabled:
> 
>     * `vlen = 128 & elen = 64`
> 
>     * `vlen = 256 & elen = 64`
> 
>     * `vlen = 512 & elen = 64`
> 
> 
> To compile SIMDe with support for the conversion from NEON to the RISC-V Vector Extension, please use Clang-17 and include the flag `-mrvv-vector-bits=<vector_length_of_vector_machine>` during compilation. Replace <vector_length_of_vector_machine> with the actual vector length of RISC-V vector machine.


https://github.com/simd-everywhere/simde/pull/1130

**Comment by howjmay** [comment]
If fixed size rvv, has been supported in the recent GCC/Clang, then I temporarily can't see the potential need for this feature. 

@zengdage @mr-c Do you guys know any case that we should support this feature. I think rvv still hasn't not reached the stable version yet.
If there is such a need exists, I am more than happy for the offer that @OMaghiarIMG's provides

In the meantime, I am ok for a patch to support fixed size rvv now. Just I hope it could be a build option which we can turn it off in default with macro

**Comment by OMaghiarIMG** [comment]
> In the meantime, I am ok for a patch to support fixed size rvv now. Just I hope it could be a build option which we can turn it off in default with macro

That's right, if the built option is detected the Neon types will be defined using the fixed RVV types, otherwise it will use the current sizeless types, no other changes. Just one setback, seems like the tuple types can't have a fixed size yet, opened topic at LLVM: https://github.com/llvm/llvm-project/issues/88369

**Comment by howjmay** [comment]
@OMaghiarIMG Thank you so much for the help!

**Comment by howjmay** [comment]
@OMaghiarIMG Will you do the fix? No rush, just curious.

**Comment by howjmay** [comment]
I think only the tuple types is ok. This way aligns to the neon way to implement tuples. It should be easy for other developers to understand the codebase.
